core/src/test/java/org/apache/iceberg/util/TestSnapshotUtil.java

```diff
@@ -101,54 +116,72 @@ public void cleanupTables() {
+    expectedSnapshots(new long[] {snapshotMain2Id, snapshotMain1Id, snapshotBaseId}, snapshots);
 
     List<Long> snapshotList = SnapshotUtil.currentAncestorIds(table);
     Assert.assertArrayEquals(
-        new Long[] {snapshotDId, snapshotBId, snapshotAId}, snapshotList.toArray(new Long[0]));
+        new Long[] {snapshotMain2Id, snapshotMain1Id, snapshotBaseId},
+        snapshotList.toArray(new Long[0]));
   }
 
   @Test
   public void oldestAncestor() {
     Snapshot snapshot = SnapshotUtil.oldestAncestor(table);
-    Assert.assertEquals(snapshotAId, snapshot.snapshotId());
+    Assert.assertEquals(snapshotBaseId, snapshot.snapshotId());
 
-    snapshot = SnapshotUtil.oldestAncestorOf(table, snapshotDId);
-    Assert.assertEquals(snapshotAId, snapshot.snapshotId());
+    snapshot = SnapshotUtil.oldestAncestorOf(table, snapshotMain2Id);
+    Assert.assertEquals(snapshotBaseId, snapshot.snapshotId());
 
-    snapshot = SnapshotUtil.oldestAncestorAfter(table, snapshotATimestamp + 1);
-    Assert.assertEquals(snapshotBId, snapshot.snapshotId());
+    snapshot = SnapshotUtil.oldestAncestorAfter(table, snapshotBaseTimestamp + 1);
+    Assert.assertEquals(snapshotMain1Id, snapshot.snapshotId());
   }
 
   @Test
   public void snapshotsBetween() {
     List<Long> snapshotIdsBetween =
-        SnapshotUtil.snapshotIdsBetween(table, snapshotAId, snapshotDId);
+        SnapshotUtil.snapshotIdsBetween(table, snapshotBaseId, snapshotMain2Id);
     Assert.assertArrayEquals(
-        new Long[] {snapshotDId, snapshotBId}, snapshotIdsBetween.toArray(new Long[0]));
+        new Long[] {snapshotMain2Id, snapshotMain1Id}, snapshotIdsBetween.toArray(new Long[0]));
 
     Iterable<Snapshot> ancestorsBetween =
-        SnapshotUtil.ancestorsBetween(table, snapshotDId, snapshotBId);
-    expectedSnapshots(new long[] {snapshotDId}, ancestorsBetween);
+        SnapshotUtil.ancestorsBetween(table, snapshotMain2Id, snapshotMain1Id);
+    expectedSnapshots(new long[] {snapshotMain2Id}, ancestorsBetween);
+
+    ancestorsBetween = SnapshotUtil.ancestorsBetween(table, snapshotMain2Id, snapshotBranchId);
+    expectedSnapshots(
+        new long[] {snapshotMain2Id, snapshotMain1Id, snapshotBaseId}, ancestorsBetween);
+  }
+
+  @Test
+  public void ancestorsOf() {
+    Iterable<Snapshot> snapshots = SnapshotUtil.ancestorsOf(snapshotFork2Id, table::snapshot);
+    expectedSnapshots(new long[] {snapshotFork2Id, snapshotFork1Id}, snapshots);
+
+    Iterator<Snapshot> snapshotIter = snapshots.iterator();
+    while (snapshotIter.hasNext()) {
+      snapshotIter.next();
+    }
 
-    ancestorsBetween = SnapshotUtil.ancestorsBetween(table, snapshotDId, snapshotCId);
-    expectedSnapshots(new long[] {snapshotDId, snapshotBId, snapshotAId}, ancestorsBetween);
+    // Once snapshot iterator has been exhausted, call hasNext again to make sure it is stable.
+    Assert.assertFalse(snapshotIter.hasNext());
```

Comment:
nit: `Assertions.assertThat(snapshotIter).isExhausted()`

Reply:
Fixed in follow-up commit.